[feat] adding basic geolocation

Alystrasz opened the pull request:
Using [geolocator](https://pub.dev/packages/geolocator) package, the app receives a location every ~5 seconds (if the user consents + if geolocation is enabled on the phone).

Alystrasz committed:
[chore] adding geolocator dependency

Alystrasz committed:
[chore] adding permissions for Android + iOS

Alystrasz committed:
[feat] listening to location updates

Alystrasz committed:
[refactor] now using DateTime in location entity

Alystrasz commented:
`LocationOptions(accuracy: LocationAccuracy.best, timeInterval: 5000)`
With this, we receive a location every few seconds, which might impact phone's battery: is it possible/reasonable to increase this interval?

rouvoy commented:
I believe that we can live with one location retrieved every 1 or 2 minutes. Most of the visited places will be indoor places, so we can also rely on Google Play services to locate the phone with minimum battery consumption (by using alternative location technology), if available on Flutter.

Alystrasz commented:
Geolocator package already uses Google Play services (at least on Android, the implementation being a [FusedLocationProviderClient](https://developers.google.com/android/reference/com/google/android/gms/location/FusedLocationProviderClient)), so that should do it 😃 
I'll increase the location timeout to 60 seconds.

Alystrasz committed:
[chore] receiving one location per minute (every 61 seconds exactly)

Alystrasz committed:
[chore] merge from master

Alystrasz committed:
[fix] moving location listener outside of build method

Alystrasz committed:
[chore] merge from master

m3ftah commented:
The app has asked for location permission, there is no bug.
How can we test this feature?

Alystrasz commented:
For now, the app requests a location every 60 seconds; each time a location is received, it is printed on the logs.

m3ftah commented:
Nothing is printed on the log.
I am using a Simulator.

m3ftah commented:
Have you tried to see if it works even if you close the app, or switch to another app on Android?

Alystrasz commented:
I launched the app, switched to a browser, and it seems that it gathers locations while in background (tested on Android 6) 
```
D/FlutterView(31752): Detaching from a FlutterEngine: io.flutter.embedding.engine.FlutterEngine@6f59ceb 
I/flutter (31752): new location stored: {lat: 51.0286983, lng: 2.3703206, date: 2020-05-04 12:15:15.774Z}
I/flutter (31752): new location stored: {lat: 51.0286983, lng: 2.3703206, date: 2020-05-04 12:16:15.809Z}
I/flutter (31752): new location stored: {lat: 51.0286982, lng: 2.3703206, date: 2020-05-04 12:17:15.890Z}
I/flutter (31752): new location stored: {lat: 51.0286982, lng: 2.3703206, date: 2020-05-04 12:18:15.915Z}
```

m3ftah commented:
Good, could you also check if this takes more than 15 minutes, as Android seems to idle apps if they are not on the foreground.

Alystrasz commented:
What would we do if this takes more than 15 minutes?

m3ftah commented:
You mean if it **doesn't** take more than 15 minutes?

Alystrasz commented:
Well, the more locations we have, the most accurate our results will be.
Moreover, we can't force people to activate geolocation on their phones, so receiving a new location definitely might take more than 15 minutes 😄

Alystrasz committed:
[chore] merge from master

Alystrasz committed:
[feat] adding information sheet details about geolocation

Alystrasz committed:
[feat] adding a geolocation switch in information sheet

Alystrasz committed:
[refactor] exporting information sheet to new dedicated component

Alystrasz committed:
[refactor] switch now has a state and can be switched on/off

Alystrasz committed:
[feat] geoswitch can be triggered by touching the list tile

Alystrasz committed:
[refactor] removing code duplication

Alystrasz committed:
[chore] adding permission_handler dependency

Alystrasz committed:
[feat] switch shows state of the location permission

Alystrasz committed:
[refactor] exporting all location list tile to dedicated component

Alystrasz committed:
[feat] displaying a message when trying to allow location

When asking for the location permission, a user might have selected
the "do not ask me again" option; we display a toast to warn him,
and then redirects him to the settings screen, where the location
permission can be updated.

Alystrasz committed:
[fix] package import was incorrect ("file:///D:/workspace/pandemia...")

Alystrasz committed:
[feat] displaying a message when trying to revoke location permission

Alystrasz committed:
[feat] displaying a message when location permission has been granted but location service is not running

Alystrasz committed:
[fix] at start, the app tried to insert location while database was not opened yet

Alystrasz committed:
[chore] attempting to add Podfile

Alystrasz committed:
[refactor] using background_locator instead of geolocator

Alystrasz committed:
[chore] removing useless imports

Alystrasz committed:
[chore] adding a custom application launcher

Alystrasz committed:
[refactor] location registration is now done in the location callback

Alystrasz committed:
[feat] geolocator can now be stopped

Alystrasz committed:
[fix] not launching location gathering if permissions have not been granted

Commit:
[chore] updating background_locator version to 1.1.10+1

Commit:
[chore] adding permission + receiver to restart location service on Android reboot

Commit:
[fix] notification icon now displays correctly

Commit:
[feat] adding translations for notification

Commit:
[fix] added french translations

Commit:
[fix] geolocation can be enabled on all smartphones

On phones with little screens, the information sheet would take all screen
space but not be completely readable due to the screen size.
The information sheet is now scrollable, making it accessible to all devices.

Alystrasz commented:
I need to check that locations gathering:
* is correctly done on iOS;
* [starts right after boot completion on Android](https://github.com/rekab-app/background_locator/issues/79).

Ideally, I also would wait for [places-popularity-map](https://github.com/Alystrasz/pandemia/tree/places-popularity-map) to be merged into master before merging this, to display locations as a heatmap on the app main page.

Alystrasz committed:
[fix] iOS app does not crash at start anymore

Alystrasz committed:
[chore] updating iOS project scheme

Alystrasz committed:
[feat] adding a human-readable message when allowing permissions in the phone's settings

Alystrasz committed:
[fix] adding missing iOS permissions

Alystrasz committed:
[feat] merge from master

Alystrasz committed:
[fix] adding new required parameter for BackgroundLocator.registerLocationUpdate

Alystrasz committed:
[fix] adding background notification channel name

Alystrasz committed:
[fix] open database before fetching locations

Alystrasz committed:
[feat] adding translations for location map

Alystrasz committed:
[feat] VisitedPlacesCard now have access to locations stored in database

Alystrasz committed:
[refactor] building visited places map from method

Alystrasz committed:
[feat] Locations can be converted to LatLng

Alystrasz committed:
[feat] computing heatmap if locations are stored

Alystrasz committed:
[refactor] VisitedPlacesCard is now stateful

Alystrasz committed:
[chore] updating Podfile

Commit:
[refactor] building locations from database with a factory method

Commit:
[fix] removing useless prints

Alystrasz committed:
Merge branch 'master' of https://github.com/TanguyHouppe/pandemia into TanguyHouppe-master

Alystrasz committed:
Merge branch 'TanguyHouppe-master' into geolocation

Alystrasz committed:
[fix] loading map style when building card state

Alystrasz committed:
[refactor] removing useless code

Alystrasz committed:
[refactor] computing markers in card state

Alystrasz committed:
[fix] setting component key as final

Alystrasz committed:
Merge branch 'master' into geolocation

Alystrasz committed:
[fix] opening database before inserting locations

Alystrasz committed:
[fix] reimporting markers computation code

Alystrasz committed:
[refactor] removing useless delay while computing exposition report

Alystrasz committed:
[refactor] renaming generateRandomReport to generateReport (as it generates real reports from now on)

Alystrasz committed:
[fix] registering Java GeolocatorPlugin to use it while in background

Alystrasz committed:
[fix] registering Java FlutterLocalNotificationsPlugin to send notifications while in background

Alystrasz committed:
[refactor] removing useless imports

Alystrasz committed:
[chore] updating iOS dependencies

Alystrasz committed:
[feat] registering background plugins for iOS

Alystrasz committed:
[chore] adjusting iOS location settings

Alystrasz committed:
[chore] using geocoding package instead of geolocator package

Packages developer splited geocoding into two packages to ease
their development, and the features we need are located on the
second package.

Alystrasz committed:
[fix] updating notification icon + channel

Alystrasz committed:
[refactor] updating location tracking notification channel name

Alystrasz committed:
[feat] adding translations for exposure notification

Alystrasz committed:
[refactor] passing BuildContext to geolocator for translation purposes

Alystrasz committed:
[refactor] translating location notification channel name

Alystrasz committed:
[feat] translating exposure notification

Alystrasz committed:
[feat] setting rate to 100 if it's above 100

Alystrasz committed:
[fix] adding loading text for places title

Alystrasz committed:
[feat] visitedPlacesCard works with fresh locations (24h old maximum)

Alystrasz committed:
[refactor] rewriting markers method to remove useless code

Alystrasz committed:
[fix] removing useless import

Alystrasz committed:
[feat] adding static class to compute visited places

Alystrasz committed:
[refactor] storing visited places in app model

Alystrasz committed:
[refactor] removing dataCollect class and transfering its method into VisitedPlacesComputer

Alystrasz committed:
[refactor] storing visited places markers in app model

Alystrasz committed:
[fix] only initializing visitedPlacesComputer once

Alystrasz committed:
[refactor] visited places are now computed when generating report

Alystrasz committed:
[refactor] directly getting current rate from computer while generating report

Alystrasz committed:
[refactor] removing useless code

Alystrasz committed:
[refactor] using popular times parser to get a place's popularity

Alystrasz committed:
[feat] getting live popularity stats while computing visited places by ignoring parser cache

Alystrasz committed:
[refactor] removing useless imports

Alystrasz committed:
[feat] computing visited places w/ only locations captured during current day

While computing locations and visited places to display on the
visitedPlacesCard component, which is meant to show map elements
for the present day only, we filter locations to retain those from
present day.

Alystrasz committed:
[fix] center camera on a default position when no location is available

Alystrasz committed:
[feat] adding method to check if phone is connected to internet

Alystrasz committed:
[fix] preventing report refresh if there's no internet connection

Alystrasz committed:
[fix] not updating exposition report at launch without internet connection

Alystrasz committed:
[feat] displaying error message at launch when no internet

Alystrasz committed:
[fix] adding missing play services dependency

Alystrasz committed:
build: update http dependency
